runtime/drivers/duckdb/olap.go

```diff
@@ -549,6 +609,61 @@ func (c *connection) RenameTable(ctx context.Context, oldName, newName string, v
 	return err
 }
 
+func (c *connection) execIncrementalInsert(ctx context.Context, safeName, sql string, byName bool, strategy drivers.IncrementalStrategy, uniqueKey []string) error {
+	var byNameClause string
+	if byName {
+		byNameClause = "BY NAME"
+	}
+
+	if strategy == drivers.IncrementalStrategyAppend {
+		return c.execWithLimits(ctx, &drivers.Statement{
+			Query:       fmt.Sprintf("INSERT INTO %s %s (%s\n)", safeName, byNameClause, sql),
+			Priority:    1,
+			LongRunning: true,
+		})
+	}
+
+	if strategy == drivers.IncrementalStrategyMerge {
+		// Create a temporary table with the new data
+		tmp := uuid.New().String()
+		err := c.execWithLimits(ctx, &drivers.Statement{
+			Query:       fmt.Sprintf("CREATE TEMPORARY TABLE %s AS (%s\n)", safeSQLName(tmp), sql),
+			Priority:    1,
+			LongRunning: true,
+		})
+		if err != nil {
+			return err
+		}
+
+		// Drop the rows from the target table where the unique key is present in the temporary table
+		where := ""
+		for i, key := range uniqueKey {
+			key = safeSQLName(key)
+			if i != 0 {
+				where += " AND "
+			}
+			where += fmt.Sprintf("base.%s = tmp.%s", key, key)
+		}
+		err = c.execWithLimits(ctx, &drivers.Statement{
+			Query:       fmt.Sprintf("DELETE FROM %s base WHERE EXISTS (SELECT 1 FROM %s tmp WHERE %s)", safeName, safeSQLName(tmp), where),
+			Priority:    1,
+			LongRunning: true,
+		})
+		if err != nil {
+			return err
+		}
+
+		// Insert the new data into the target table
+		return c.execWithLimits(ctx, &drivers.Statement{
```

Comment:
It's a good point and admittedly the `merge` support is somewhat experimental at this point (also not sure how it will perform). 

Doing a delete+insert is a common merge strategy for databases without native merge support. Ideally though, we would do it in a transaction, but I wonder if that might give us other problems with DuckDB.

For external table storage, I added the `inPlace == false` option where it takes a copy of the table to prevent this issue. It's used for DuckDB->DuckDB models, for unfortunately the SQLStore->DuckDB transporter currently does streaming writes, so couldn't use it there.

I'm planning more refactors here to eventually enable safe writes (among other things), but for now, this PR is growing big and would prefer to get it merged. Since the merge support is experimental and doesn't break any existing features, I think it should be fine.

If you would prefer, we could try adding a transaction here for the short term?

Reply:
(And to address the question of context specifically – it's probably best to use the context since these operations can run for a long time. And in terms of error scenarios, using a new context would not address the issue of the second query failing or the process being terminated.)

Reply:
Yeah given its not an issue for DuckDB->DuckDB models which are also going to be the most common ones we can handle `merge` in later PRs.